lib/src/rules/prefer_collection_literals.dart

```diff
@@ -142,35 +142,35 @@ class _Visitor extends SimpleAstVisitor<void> {
 
   bool _shouldSkipLinkedHashLint(
       InstanceCreationExpression node, bool Function(DartType node) typeCheck) {
-    if (_isHashMap(node) || _isHashSet(node)) {
-      // Skip: LinkedHashSet<int> s =  ...; or LinkedHashMap<int> s =  ...;
-      var parent = node.parent;
-      if (parent is VariableDeclaration) {
-        var parent2 = parent.parent;
-        if (parent2 is VariableDeclarationList) {
-          var assignmentType = parent2.type?.type;
-          if (assignmentType != null && !typeCheck(assignmentType)) {
-            return true;
-          }
-        }
-      }
-      // Skip: function(LinkedHashSet()); when function(LinkedHashSet mySet) or
-      // function(LinkedHashMap()); when function(LinkedHashMap myMap)
-      if (parent is ArgumentList) {
-        final paramType = parent.arguments.first.staticParameterElement?.type;
-        if (paramType != null && !typeCheck(paramType)) {
-          return true;
-        }
-      }
-      // Skip: <int, LinkedHashSet>{}.putIfAbsent(3, () => LinkedHashSet());
-      // or <int, LinkedHashMap>{}.putIfAbsent(3, () => LinkedHashMap());
-      if (parent is ExpressionFunctionBody) {
-        var expressionType = parent.expression.staticType;
-        if (expressionType != null && !typeCheck(expressionType)) {
-          return true;
-        }
-      }
+    if (!_isHashMap(node) && !_isHashSet(node)) return false;
+    var parent = node.parent;
+    var staticType = _enforcedType(parent);
+    return staticType != null && !typeCheck(staticType);
+  }
+}
+
+/// Returns the static type which is pushed into an expression by it's parent.
+DartType /*?*/ _enforcedType(AstNode parent) {
+  if (parent is VariableDeclaration) {
+    var parent2 = parent.parent;
+    if (parent2 is VariableDeclarationList) {
+      return parent2.type?.type;
     }
-    return false;
   }
+  if (parent is ArgumentList) {
+    return parent.arguments.first.staticParameterElement?.type;
+  }
+  if (parent is NamedExpression) {
+    return parent.staticType;
+  }
+  if (parent is ExpressionFunctionBody) {
+    return parent.expression.staticType;
```

Comment:
This also looks wrong. `parent.expression` will always return the instance creation expression from which we started. I think what you want to do here is recurse to see whether you can find a type that would be inferred as the return type of the closure (assuming that this is the body of a closure).

Reply:
This is another existing bug with false negatives.

Is this something you might be able to work on? I can easily add some test cases showing the problem but I'm having a hard time figuring out the right things to look for in the analyzer APIs.

Reply:
If you can send me the test case(s), I'll be happy to at least look at what it would take to implement them.

Reply:
https://github.com/dart-lang/linter/pull/3705/files adds a line in the test data library that should have a lint but doesn't due to this bug.